pkg/processor/build/runtime/suite/suite.go

```diff
@@ -0,0 +1,158 @@
+/*
+Copyright 2017 The Nuclio Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package runtimesuite
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/nuclio/nuclio/pkg/dockerclient"
+	"github.com/nuclio/nuclio/pkg/processor/build"
+	"github.com/nuclio/nuclio/pkg/zap"
+
+	"github.com/nuclio/nuclio-sdk"
+	"github.com/rs/xid"
+	"github.com/stretchr/testify/suite"
+)
+
+//
+// Base suite
+//
+
+type RuntimeTestSuite struct {
+	suite.Suite
+	Logger       nuclio.Logger
+	Builder      *build.Builder
+	DockerClient *dockerclient.Client
+	TestID       string
+}
+
+func (suite *RuntimeTestSuite) SetupSuite() {
+	var err error
+
+	suite.Logger, err = nucliozap.NewNuclioZapTest("test")
+	suite.Require().NoError(err)
+
+	suite.DockerClient, err = dockerclient.NewClient(suite.Logger)
+	suite.Require().NoError(err)
+}
+
+func (suite *RuntimeTestSuite) SetupTest() {
+	suite.TestID = xid.New().String()
+}
+
+func (suite *RuntimeTestSuite) BuildAndRunFunction(functionName string,
+	functionPath string,
+	runtime string,
+	ports map[int]int,
+	requestPort int,
+	requestBody string,
+	expectedResponseBody string) {
+
+	var err error
+
+	functionName = fmt.Sprintf("%s-%s", functionName, suite.TestID)
+	imageName := fmt.Sprintf("nuclio/processor-%s", functionName)
+
+	suite.Builder, err = build.NewBuilder(suite.Logger, &build.Options{
+		FunctionName:    functionName,
+		FunctionPath:    functionPath,
+		Runtime:         runtime,
+		NuclioSourceDir: suite.GetNuclioSourceDir(),
+		Verbose:         true,
+	})
+
+	suite.Require().NoError(err)
+
+	// do the build
+	err = suite.Builder.Build()
+	suite.Require().NoError(err)
+
+	// remove the image when we're done
+	defer suite.DockerClient.RemoveImage(imageName)
+
+	// run the processor
+	containerID, err := suite.DockerClient.RunContainer(imageName, ports, "")
+
+	suite.Require().NoError(err)
+
+	// remove the container when we're done
+	defer suite.DockerClient.RemoveContainer(containerID)
+
+	// give the container some time - after 10 seconds, give up
+	deadline := time.Now().Add(10 * time.Second)
+
+	for {
```

Comment:
IMO it'll be cleaner if processor will have a heartbeat API and you'll be able to have `WaitForProcessor(timeout time.Duration)` function using the usual `select` and `time.After`.

Reply:
IMO generic heartbeats never work in the real world. Your service might not be listening yet / not serving its purpose yet your heartbeat works